src/components/views/rooms/AudioRecorder.js

```diff
@@ -0,0 +1,111 @@
+//@flow
+
+import React from 'react';
+import classNames from 'classnames';
+
+import sdk from '../../../index';
+import Modal from '../../../Modal';
+import ErrorDialog from '../dialogs/ErrorDialog';
+
+export default class AudioRecorder extends React.Component {
+  static propTypes = {
+    sendAudioBlob: React.PropTypes.func.isRequired,
+  };
+
+  mediaRecorder: ?MediaRecorder = null;
+  blobs: Array<Blob> = [];
+
+  constructor(props: Object) {
+    super(props);
+
+    this.state = {
+      recording: false,
+    };
+  }
+
+  onRecordClicked = async () => {
+    if(this.state.recording) {
+      this.mediaRecorder.stop();
+      this.mediaRecorder = null;
+      this.mediaStream.getTracks().forEach((track) => track.stop());
+      this.mediaStream = null;
+    } else {
+      if(!navigator.mediaDevices || !MediaRecorder) {
+        Modal.createDialog(ErrorDialog, {
+          description: `Sorry, but the features required for push 
+ to talk messages are not supported by your browser`,
+        });
+        return;
+      }
+
+      try {
+        const mediaStream = await navigator.mediaDevices.getUserMedia({
+          audio: true,
+          video: false,
+        });
+        this.onUserMediaSuccess(mediaStream);
+      } catch (e) {
+        this.onUserMediaFailure(e);
+      }
+    }
+  };
+
+  onUserMediaFailure = () => {
+    Modal.createDialog(ErrorDialog, {
+      description: "Failed to start recording.",
+    });
+  };
+
+  onUserMediaSuccess = (mediaStream: MediaStream) => {
+    this.mediaStream = mediaStream;
+
+    this.mediaRecorder = new MediaRecorder(mediaStream);
```

Comment:
Leaving out the options parameter, in particular `mimeType`, seems unwise, given the default is likely to vary between browsers. What format are we expecting here? Bear in mind we'll need to spec this so it can be implemented cross-client, we'll need to nail down what audio formats clients should expect.

Reply:
There is the option to check for supported mimeTypes: MediaRecorder.isTypeSupported(). Expected is "audio/webm\;codecs=opus" for Chrome and "audio/ogg\;codecs=opus" for Firefox. (Keep in mind that Safari doesn't support MediaRecorder.) Seems like Firefox can only record ogg audio and Chrome only WebM audio (see https://stackoverflow.com/questions/35466078/specifying-codecs-with-mediarecorder). Both browser can play both file types though, so that's not a problem.